pkg/processor/runtime/rpc/connection/socketallocator.go

```diff
@@ -0,0 +1,218 @@
+
+	"github.com/nuclio/nuclio/pkg/common"
+
+	"github.com/nuclio/errors"
+	"github.com/rs/xid"
+)
+
+const (
+	socketPathTemplate = "/tmp/nuclio-rpc-%s.sock"
+	connectionTimeout  = 2 * time.Minute
+)
+
+type SocketAllocator struct {
+	*AbstractConnectionManager
+
+	eventSockets         []*EventSocket
+	controlMessageSocket *ControlMessageSocket
+}
+
+func NewSocketAllocator(baseConnectionManager *AbstractConnectionManager) *SocketAllocator {
+	return &SocketAllocator{
+		AbstractConnectionManager: baseConnectionManager,
+		eventSockets:              make([]*EventSocket, 0),
+	}
+}
+
+func (sa *SocketAllocator) Prepare() error {
+	if sa.Configuration.SupportControlCommunication {
+		controlConnection, err := sa.createSocketConnection()
+		if err != nil {
+			return errors.Wrap(err, "Failed to create socket connection")
+		}
+		sa.controlMessageSocket = NewControlMessageSocket(
+			sa.Logger,
+			controlConnection,
+			sa.RuntimeConfiguration.ControlMessageBroker)
+	}
+
+	for i := 0; i < sa.MinSocketsNum; i++ {
+		eventConnection, err := sa.createSocketConnection()
+		if err != nil {
+			return errors.Wrap(err, "Failed to create socket connection")
+		}
+		sa.eventSockets = append(sa.eventSockets, NewEventSocket(sa.Logger,
+			eventConnection, sa))
+	}
+	return nil
+}
+
+func (sa *SocketAllocator) Start() error {
+	if err := sa.startSockets(); err != nil {
+		return errors.Wrap(err, "Failed to start socket allocator")
+	}
+
+	// wait for start if required to
+	if sa.Configuration.WaitForStart {
+		sa.Logger.Debug("Waiting for start")
+		for _, socket := range sa.eventSockets {
+			socket.Start()
+		}
+	}
+
+	sa.Logger.Debug("Socker allocator started")
+	return nil
+}
+
+func (sa *SocketAllocator) Stop() error {
+	for _, eventSocket := range sa.eventSockets {
+		socket := eventSocket
+		go func() {
+			socket.Stop()
+		}()
+	}
+	if sa.controlMessageSocket != nil {
+		go func() {
+			sa.controlMessageSocket.Stop()
+		}()
+	}
+	return nil
+}
+
+func (sa *SocketAllocator) Allocate() (EventConnection, error) {
+	// TODO: implement allocation logic when support multiple sockets
+	return sa.eventSockets[0], nil
+}
+
+func (sa *SocketAllocator) GetAddressesForWrapperStart() ([]string, string) {
+	eventAddresses := make([]string, 0)
+
+	for _, socket := range sa.eventSockets {
+		eventAddresses = append(eventAddresses, socket.Address)
+	}
+
+	if sa.controlMessageSocket == nil {
+		sa.Logger.DebugWith("Get socket addresses",
+			"eventAddresses", eventAddresses,
+			"controlAddress", "")
+		return eventAddresses, ""
+	}
+	sa.Logger.DebugWith("Get socket addresses",
+		"eventAddresses", eventAddresses,
+		"controlAddress", sa.controlMessageSocket.Address)
+	return eventAddresses, sa.controlMessageSocket.Address
+}
+
+func (sa *SocketAllocator) startSockets() error {
+	var err error
+	for _, socket := range sa.eventSockets {
+		if socket.Conn, err = socket.listener.Accept(); err != nil {
+			return errors.Wrap(err, "Can't get connection from wrapper")
+		}
```

Comment:
if we have some sockets that succeeded and one that failed, what happens to the successful ones?

Reply:
@TomerShor then we fail, I'll add TODO here to come back to this question when we support more than 1 socket